wsrep_listener.c

```diff
@@ -55,27 +55,33 @@ view_cb (void*                    app_ctx   __attribute__((unused)),
          void*                    recv_ctx  __attribute__((unused)),
          const wsrep_view_info_t* view,
          const char*              state     __attribute__((unused)),
-         size_t                   state_len __attribute__((unused)),
-         void**                   sst_req,
-         size_t*                  sst_req_len)
+         size_t                   state_len __attribute__((unused)))
 {
     printf ("New cluster membership view: %d nodes, my index is %d, "
             "global seqno: %lld\n",
             view->memb_num, view->my_idx, (long long)view->state_id.seqno);
 
-    if (view->state_gap) /* we need to receive new state from the cluster */
-    {
-        /* For simplicity we're skipping state transfer by using magic string
-         * as a state transfer request.
-         * This node will not be considered JOINED (having full state)
-         * by other cluster members. */
-        *sst_req = strdup(WSREP_STATE_TRANSFER_NONE);
-
-        if (*sst_req)
-            *sst_req_len = strlen(*sst_req) + 1;
-        else
-            *sst_req_len = -ENOMEM;
-    }
+    return WSREP_CB_SUCCESS;
+}
+
+/*! This will be called on cluster view change (nodes joining, leaving, etc.).
+ *  Each view change is the point where application may be pronounced out of
+ *  sync with the current cluster view and need state transfer.
+ *  It is guaranteed that no other callbacks are called concurrently with it. */
+static wsrep_cb_status_t
+sst_request_cb (void**            sst_req,
+                size_t*           sst_req_len)
+{
+    /* For simplicity we're skipping state transfer by using magic string
+     * as a state transfer request.
+     * This node will not be considered JOINED (having full state)
+     * by other cluster members. */
+    *sst_req = strdup(WSREP_STATE_TRANSFER_NONE);
+
+    if (*sst_req)
+        *sst_req_len = strlen(*sst_req) + 1;
+    else
+        *sst_req_len = -ENOMEM;
```

Comment:
If ```sst_req_len``` stores negative values for errors, the sst_req_len type should be ```ssize_t```. Or return ```wsrep_cb_status_t``` error code in case of failure and leave ```sst_req_len``` untouched (now the function returns success even if ```strdup()``` fails).